arch/x86/adler32_sse41.c

```diff
@@ -0,0 +1,101 @@
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#include "../../zbuild.h"
+#include "../../zutil.h"
+
+#include "../../adler32_p.h"
+
+#ifdef X86_SSE41_ADLER32
+
+#include <immintrin.h>
+
+static inline uint32_t partial_hsum(__m128i x) {
+    __m128i second_int = _mm_bsrli_si128(x, 8);
+    __m128i sum = _mm_add_epi32(x, second_int);
+    return _mm_cvtsi128_si32(sum);
+}
+
+static inline uint32_t hsum(__m128i x) {
+    __m128i sum1 = _mm_unpackhi_epi64(x, x);
+    __m128i sum2 = _mm_add_epi32(x, sum1);
+    __m128i sum3 = _mm_shuffle_epi32(sum2, 0x01);
+    __m128i sum4 = _mm_add_epi32(sum2, sum3);
+    return _mm_cvtsi128_si32(sum4);
+}
+
+Z_INTERNAL uint32_t adler32_sse41(uint32_t adler, const unsigned char *buf, size_t len) {
+    uint32_t sum2;
+
+     /* split Adler-32 into component sums */
+    sum2 = (adler >> 16) & 0xffff;
+    adler &= 0xffff;
+
+    /* in case user likes doing a byte at a time, keep it fast */
+    if (UNLIKELY(len == 1))
+        return adler32_len_1(adler, buf, sum2);
+
+    /* initial Adler-32 value (deferred check for len == 1 speed) */
+    if (UNLIKELY(buf == NULL))
+        return 1L;
+
+    /* in case short lengths are provided, keep it somewhat fast */
+    if (UNLIKELY(len < 16))
+        return adler32_len_16(adler, buf, len, sum2);
+
+    const __m128i dot2v = _mm_setr_epi8(16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1);
+    const __m128i dot3v = _mm_set1_epi16(1);
+    const __m128i zero = _mm_setzero_si128();
+
+    __m128i vs1 = _mm_cvtsi32_si128(adler);
+    __m128i vs2 = _mm_insert_epi32(_mm_setzero_si128(), sum2, 3);
+
+    while (len >= 16) {
+       __m128i vs1_0 = vs1;
+       __m128i vs3 = _mm_setzero_si128();
+
+       int k = (len < NMAX ? (int)len : NMAX);
+       k -= k % 16;
+       len -= k;
+
+       while (k >= 16) {
+           /*
+              vs1 = adler + sum(c[i])
+              vs2 = sum2 + 16 vs1 + sum( (16-i+1) c[i] )
+           */
+           __m128i vbuf = _mm_loadu_si128((__m128i*)buf);
+           buf += 16;
+           k -= 16;
+
+           __m128i v_sad_sum1 = _mm_sad_epu8(vbuf, zero);
+           vs1 = _mm_add_epi32(v_sad_sum1, vs1);
+           vs3 = _mm_add_epi32(vs1_0, vs3);
+           __m128i v_short_sum2 = _mm_maddubs_epi16(vbuf, dot2v);
+           __m128i vsum2 = _mm_madd_epi16(v_short_sum2, dot3v);
+           vs2 = _mm_add_epi32(vsum2, vs2);
+           vs1_0 = vs1;
+       }
+
+       vs3 = _mm_slli_epi32(vs3, 4);
+       vs2 = _mm_add_epi32(vs2, vs3);
+
+       /* We don't actually need to do a full horizontal sum, since psadbw is actually doing
+        * a partial reduction sum implicitly and only summing to integers in vector positions
+        * 0 and 2. This saves us some contention on the shuffle port(s) */
+       adler = partial_hsum(vs1) % BASE;
+       sum2 = hsum(vs2) % BASE;
+
+       vs1 = _mm_cvtsi32_si128(adler);
+       vs2 = _mm_insert_epi32(_mm_setzero_si128(), sum2, 3);
```

Comment:
Why are these using different intrinsics even though both move 32-bit integer to vector register at same position.

Reply:
They don't.  The top one moves it at the first position (first 4 bytes in memory with little endian).  The other moves things to the last 4 bytes.  This is needed so that the multiplication done in the madd lines up.

Reply:
> README.md also needs to be updated in two places.

will do

Reply:
As this differs from other adler32 versions, it should be properly commented...

Reply:
Hmm, don't know why github's quote reply somehow got attached to this thread...

Reply:
I've always thought `_mm_insert_epi*` instructions are expensive instructions, basically taking 2 cycles, combined with 0.33 cycles for _mm_setzero_si128(), which also requires own register as it's basically implemented as xor with itself. Both `_mm_cvtsi32_si128` and shifting the vector contents take 1 cycles, which should save 0.33 cycles and one vector register.

Reply:
Well there's a few things happening here. The xor to set to zero is cheap and we can do a lot of them at once. But additionally, it's a zeroing idiom which may in fact make the register rename engine in the register file just quickly swap in a zero'd register. In addition to that, I don't think that the movd instruction is one cycle. According to agner fog's instruction table, on wolfdale movd between GPR and XMM register is 2 cycles.

Reply:
Also the shifting of bytes still requires the load to complete. The best we could hope for is that it'd overlap the movd that precedes it. But that shift puts us outside of that 2 cycle range for overlap.

Reply:
And also the same seems to be true for Nehalem as well:
https://www.agner.org/optimize/instruction_tables.pdf
Though, he qualifies the latency a little better in those tables.  The latency is 1+1, where there's a bypass delay that has to occur when crossing the GPR/XMM boundary.

Reply:
> As this differs from other adler32 versions, it should be properly commented...

I actually don't know that it necessarily needs to differ from the other implementations.  The front half of the checksum never actually required the existing running sum to be in any particular position in the vector register.  It was only the second half (sum2) that needed this.  We could just as easily modify the ssse3 variant in this commit to do the same exact thing.

Reply:
I don't really trust any tables listing instruction latencies and throughputs as the real-life values highly depend on how the processor and compiler reorder the instructions to utilize as many concurrent ports as possible.

In both cases it has to move data from GPR to xmm register, which means there is delay when it crosses from one execution unit to another. Usually the microcode is optimized better when moving to vector element 0, otherwise it wouldn't make sense to have separate instruction mnemonic for that.

It would be ideal if we could avoid trip to GPR register during the horizontal sum, but that is unlikely to be possible with any of the current processors as modulo is really expensive operation using vector registers.

Reply:
Generally speaking these values are found with some pretty careful experimentation, but I think Intel's documentation will probably confirm the same.
In any case, even though on newer architectures (as in, ivy bridge and newer), it's only 1 cycle to do the GPR -> XMM move (as documented via Intel's own intrinsic guide, anyway), you still have to wait for the move to complete before doing the byte shift, meaning this would take net 2 cycles. 

Now where this might make a difference is if you're dealing with ymm registers, as _those_ generally need to expand to a pinsert instruction + a combination of 2 XMM registers.  For that, it makes way more sense to do a broadcast + and with a static mask ymm register, as that is likely to be the cheapest way to do it.

Reply:
Hah, actually...it's not even important for us to put sum2 in the last vector position.  I'm not exactly sure why any of this code was doing this, as it just needs to be summed, not multiplied at all.  Interesting...this certainly resolves this discussion, at least.

Reply:
Well that's rich...I modified the AVX512 variant to use the two simpler vmovd instructions to xmm registers, allowing the register aliasing to zero the other bits when it turns into a zmm register and...it was slower.  I think there's a penalty when doing that, but the weird part is we should only be seeing at first register access which would be at the top of the loop and every time the loop did horizontal sums to prevent overflow (right at NMAX).  Either that, or we're able to do things on more execution ports somehow by mixing the vbroadcastd (with mask) with the vmovd instruction.  

It's weird because the compiler does this trick all the time with xmm registers.  It often will pxor an xmm register for zero before it gets used as a ymm register.  I guess I just leave the avx512 version as-is for now, populating sum2 in the last slot for now (using the first slot turns on an optimization that makes the compiler use vmovd instead).

Reply:
So, it definitely doesn't seem to be an instruction alignment thing.  Inserting two NOPs after the vmovd's to give the same lengths (while keeping the bottom of the loop where vs1 and vs2 are reset) doesn't seem to do anything for speed on the avx512 vnni version.  We must be hitting some sort of transition penalty from xmm->zmm when both vs1 and vs2 are set as xmm registers.